dask_expr/_expr.py

```diff
@@ -311,25 +317,79 @@ def rewrite(self, kind: str):
 
         return expr
 
-    def simplify(self):
+    def simplify_once(self, dependents: defaultdict):
```

Comment:
I'm a little worried about runtime complexity here. I know that expr graphs are supposed to be small but what I see is definitely a non-linear runtime in the number of edges `E` (at the least)

Generally, this function is called more than once per node. In fact, due to line 365, this is called `E` times. 
`_simplify_up` in L350 then accepts the dependents and may call things like `determine_column_projection` which iterates over the dependents which effectively brings us to a quadratic runtime in `E`

Reply:
This is not necessarily a showstopper but it is concerning, paritcularly if `simplify_up` became more complex down the road.

Reply:
`substitute_parameters` which is also called occasionally in simplify_up is also iterating over the edges so that's a second source besides `determine_column_projection` for this runtime complexity.

Reply:
@fjetter and I chatted offline a little about this

It's a potential issue for us that I intend to investigate in January. We can avoid some of this by caching the projected columns on every simplify step, which would hopefully solve the runtime complexity issue. Current main exhibits the same pattern though, so this does not block this pr

Reply:
@fjetter - Do you think this complexity may have contributed to https://github.com/dask-contrib/dask-expr/issues/796? Sorry - I never got a chance to look through these changes, so I'm not really sure how the new system works yet.

Reply:
> Do you think this complexity may have contributed to

Possibly. I'll try to have a look at this shortly